qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -151,14 +178,17 @@ def _decomposer_2q_from_basis_gates(basis_gates, pulse_optimize=None, approximat
     kak_gate = _choose_kak_gate(basis_gates)
     euler_basis = _choose_euler_basis(basis_gates)
     basis_fidelity = approximation_degree or 1.0
-    if isinstance(kak_gate, RZXGate):
-        backup_optimizer = TwoQubitBasisDecomposer(
-            CXGate(),
-            basis_fidelity=basis_fidelity,
-            euler_basis=euler_basis,
-            pulse_optimize=pulse_optimize,
-        )
-        decomposer2q = XXDecomposer(euler_basis=euler_basis, backup_optimizer=backup_optimizer)
+    # if isinstance(kak_gate, RZXGate):
+    #    backup_optimizer = TwoQubitBasisDecomposer(
+    #        CXGate(),
+    #        basis_fidelity=basis_fidelity,
+    #        euler_basis=euler_basis,
+    #        pulse_optimize=pulse_optimize,
+    #    )
+    #    decomposer2q = XXDecomposer(euler_basis=euler_basis, backup_optimizer=backup_optimizer)
+
+    if kak_gate in KAK_GATE_PARAM_NAMES.values():
+        decomposer2q = TwoQubitControlledUDecomposer(kak_gate)
```

Comment:
Double checking our discussion from earlier in the week: for the target-aware code we discussed adding a third check (apart from `is_controlled` or `is_supercontrolled`) to differentiate cases for the 3 available decomposers: `TwoQubitDecomposer`, `XXDecomposer`, and `TwoQubitControlledUDecomposer`. However, for the Python code (just basis gates), it looks like we are simply substituting the `XXDecomposer` for the `TwoQubitControlledUDecomposer`. Is this the plan?

Reply:
In the case of basis gates, I'm not sure if we can have 3 options or only 2 (this is why this code is only commented out and not removed yet).
If a user provides the basis gates `[rzz, rz, ry]` how can we tell if they want an `RZZGate` with an arbitrary angle or an `RZZGate` with a fixed angle?

Reply:
Another comment: I ran the tests in `qiskit/test/python/transpiler/test_unitary_synthesis.py` (main branch), and couldn't find any test that goes into the following "if" statement (lines 181-188 above), so I wonder if the `XXDecomposer` here has ever been used.

Reply:
I am not surprised, the coverage of these tests is not good because a lot of the tests are in the synthesis folder and not testing the transpilation pipeline, but calling the synthesis methods directly. Maybe this is a good opportunity to do some test refactoring before merging this PR. During the Rust migration I also noticed there weren't any tests for the `UnitarySynthesis` class that used the QSD method and had to add a couple of them. I think it would make sense to ensure that we have a couple of tests per decomposer, and that there are sufficient tests for the target (HW-aware) path (this I could do in a separate PR because I didn't have time to look more into it during the Rust migration).

Reply:
> If a user provides the basis gates [rzz, rz, ry] how can we tell if they want an RZZGate with an arbitrary angle or an RZZGate with a fixed angle?

I think this is up to us to decide and document the behavior change. I am not sure how many users actually rely or expect the fixed angle?

Reply:
FYI, I started the test cleanup with this PR: https://github.com/Qiskit/qiskit/pull/13582, that is not much but I think can be helpful for future development. Let me know if you have any thoughts on it!

Reply:
Another question is what to expect when given a parametrized gate in the target? see e.g this [test](https://github.com/Qiskit/qiskit/blob/079f69f7dd906299dfdf8c1262fa9bc1d75e478b/test/python/transpiler/test_unitary_synthesis.py#L1026).

Currently, we replace the parametrized gates by gates with a constant angle, without even checking if this angle is actually calibrated in the target:
https://github.com/Qiskit/qiskit/blob/079f69f7dd906299dfdf8c1262fa9bc1d75e478b/qiskit/transpiler/passes/synthesis/unitary_synthesis.py#L801

Again, I could not find any test that checks the `XXDecomposer` with a given target, namely the tests don't go into [here](https://github.com/Qiskit/qiskit/blob/079f69f7dd906299dfdf8c1262fa9bc1d75e478b/qiskit/transpiler/passes/synthesis/unitary_synthesis.py#L941).